src/core/recv_buffer.c

```diff
@@ -164,14 +181,66 @@ QuicRecvBufferHasUnreadData(
 }
 
 _IRQL_requires_max_(DISPATCH_LEVEL)
-void
-QuicRecvBufferIncreaseVirtualBufferLength(
+BOOLEAN
+QuicRecvBufferTryIncreaseVirtualBufferLength(
     _In_ QUIC_RECV_BUFFER* RecvBuffer,
     _In_ uint32_t NewLength
     )
 {
+    if (RecvBuffer->RecvMode == QUIC_RECV_BUF_MODE_EXTERNAL) {
+        return FALSE;
+    }
+
     CXPLAT_DBG_ASSERT(NewLength >= RecvBuffer->VirtualBufferLength); // Don't support decrease.
     RecvBuffer->VirtualBufferLength = NewLength;
+    return TRUE;
+}
+
+_IRQL_requires_max_(DISPATCH_LEVEL)
+QUIC_STATUS
+QuicRecvBufferProvideChunks(
+    _Inout_ QUIC_RECV_BUFFER* RecvBuffer,
+    _Inout_ CXPLAT_LIST_ENTRY* /* QUIC_RECV_CHUNKS */ Chunks
+    )
+{
+    CXPLAT_DBG_ASSERT(RecvBuffer->RecvMode == QUIC_RECV_BUF_MODE_EXTERNAL);
+    CXPLAT_DBG_ASSERT(!CxPlatListIsEmpty(Chunks));
+
+    uint64_t NewBufferLength = RecvBuffer->VirtualBufferLength;
+    for (CXPLAT_LIST_ENTRY* Link = Chunks->Flink;
+         Link != Chunks;
+         Link = Link->Flink) {
+        QUIC_RECV_CHUNK* Chunk = CXPLAT_CONTAINING_RECORD(Link, QUIC_RECV_CHUNK, Link);
+        NewBufferLength += Chunk->AllocLength;
+        if (NewBufferLength < Chunk->AllocLength) {
+            //
+            // Overflow: we can't handle that much buffer space.
+            //
+            return QUIC_STATUS_INVALID_PARAMETER;
+        }
```

Comment:
Why are you limiting the chunks to less than the virtual length, instead of just increasing the length? Also, if we keep it, I'd just optimize this to a single check after summing everything instead of doing a check for each chunk.

Reply:
This was checking for overflow of the addition above. But thinking about it, since we run in the app process and can trust it to respect the API contract, we don't need that check, overflowing the uint64 would mean allocating more memory than reasonably possible. I'll remove this check.